assessment-api/assessment-actors/src/main/scala/org/sunbird/managers/CopyManager.scala

```diff
@@ -0,0 +1,409 @@
+        case AssessmentConstants.QUESTION_MIME_TYPE =>
+          node.setInRelations(null)
+          copyNode(node, request)
+      }
+      copiedNodeFuture.map(copiedNode => {
+        val response = ResponseHandler.OK()
+        response.put("node_id", new util.HashMap[String, AnyRef]() {
+          {
+            put(node.getIdentifier, copiedNode.getIdentifier)
+          }
+        })
+        response.put(AssessmentConstants.VERSION_KEY, copiedNode.getMetadata.get(AssessmentConstants.VERSION_KEY))
+        response
+      })
+    }).flatMap(f => f) recoverWith { case e: CompletionException => throw e.getCause }
+  }
+
+  def validateExistingNode(request: Request, node: Node) = {
+    val requestObjectType = request.getObjectType
+    val nodeObjectType = node.getObjectType
+    if (!StringUtils.equalsIgnoreCase(requestObjectType, nodeObjectType)) throw new ClientException(AssessmentConstants.ERR_INVALID_OBJECT_TYPE, s"Please Provide Valid ${requestObjectType} Identifier")
+    if (StringUtils.equalsIgnoreCase(node.getObjectType, AssessmentConstants.QUESTION) && StringUtils.equalsIgnoreCase(node.getMetadata.getOrDefault(AssessmentConstants.VISIBILITY, AssessmentConstants.VISIBILITY_PARENT).asInstanceOf[String], AssessmentConstants.VISIBILITY_PARENT))
+      throw new ClientException(AssessmentConstants.ERR_INVALID_REQUEST, "Question With Visibility Parent Cannot Be Copied Individually!")
+  }
+
+  def copyQuestionSet(originNode: Node, request: Request)(implicit ex: ExecutionContext, oec: OntologyEngineContext): Future[Node] = {
+    val copyType = request.getRequest.get(AssessmentConstants.COPY_TYPE).asInstanceOf[String]
+    copyNode(originNode, request).map(node => {
+      val req = new Request(request)
+      req.put(AssessmentConstants.ROOT_ID, request.get(AssessmentConstants.IDENTIFIER))
+      req.put(AssessmentConstants.MODE, request.get(AssessmentConstants.MODE))
+      HierarchyManager.getHierarchy(req).map(response => {
+        val originHierarchy = response.getResult.getOrDefault(AssessmentConstants.QUESTIONSET, new util.HashMap[String, AnyRef]()).asInstanceOf[java.util.Map[String, AnyRef]]
+        copyType match {
+          case AssessmentConstants.COPY_TYPE_SHALLOW => updateShallowHierarchy(request, node, originNode, originHierarchy)
+          case _ => updateHierarchy(request, node, originNode, originHierarchy, copyType)
+        }
+      }).flatMap(f => f)
+    }).flatMap(f => f) recoverWith { case e: CompletionException => throw e.getCause }
+  }
+
+
+  def copyNode(node: Node, request: Request)(implicit ec: ExecutionContext, oec: OntologyEngineContext): Future[Node] = {
+    val copyCreateReq: Future[Request] = getCopyRequest(node, request)
+    copyCreateReq.map(req => {
+      DataNode.create(req).map(copiedNode => {
+        Future(copiedNode)
+      }).flatMap(f => f)
+    }).flatMap(f => f)
+  }
+
+
+  def generateNodeBLRecord(nodesModified: util.HashMap[String, AnyRef]): util.HashMap[String, AnyRef] = {
```

Comment:
Please move out all methods related to branchingLogic and put it in BranchingUtil file